hdk/virtual_machine/code.py

```diff
@@ -0,0 +1,171 @@
+"""Provides functions to translate Instruction into their assembly commands."""
+from collections.abc import Iterable, Iterator
+
+from hdk.virtual_machine.syntax import (
+    ArithmeticLogicalCommand,
+    Instruction,
+    MemoryTransferCommand,
+)
+
+_BINARY_TABLE = {
+    "add": "M=D+M",
+    "sub": "M=M-D",
+    "and": "M=D&M",
+    "or": "M=D|M",
+}
+
+_COMPARISON_TABLE = {
+    "eq": "JEQ",
+    "lt": "JLT",
+    "gt": "JGT",
+}
+
+_SEGMENT_TABLE = {
+    "local": "LCL",
+    "argument": "ARG",
+    "this": "THIS",
+    "that": "THAT",
+}
+
+
+def translate_stack_instruction(
+    stack_instruction: ArithmeticLogicalCommand, instruction_counter: int
+) -> tuple[list[str], int]:
+    """Translates a StackInstruction into its assembly code.
+
+    Args:
+        stack_instruction (StackInstruction): The stack instruction to translate.
+        instruction_counter (int): The current instruction counter value.
+
+    Returns:
+         a list of assembly code instructions and the updated instruction counter value.
+
+    Raises:
+        ValueError: If the given StackInstruction has an invalid command.
+    """
+    if stack_instruction.command == "neg":
+        return ["@SP", "A=M-1", "M=-M"], instruction_counter
+    elif stack_instruction.command == "not":
+        return ["@SP", "A=M-1", "M=!M"], instruction_counter
+    elif stack_instruction.command in _BINARY_TABLE:
+        return [
+            "@SP",
+            "AM=M-1",
+            "D=M",
+            "A=A-1",
+            _BINARY_TABLE[stack_instruction.command],
+        ], instruction_counter
+    elif stack_instruction.command in {"eq", "gt", "lt"}:
+        true_jump = _COMPARISON_TABLE[stack_instruction.command]
+        code = [
+            "@SP",
+            "AM=M-1",
+            "D=M",
+            "A=A-1",
+            "MD=M-D",
+            f"@true{instruction_counter}",
+            "D;" + true_jump,
+            f"(false{instruction_counter})",
+            "@SP",
+            "A=M-1",
+            "M=0",
+            f"@continue{instruction_counter}",
+            "0;JMP",
+            f"(true{instruction_counter})",
+            "@SP",
+            "A=M-1",
+            "M=-1",
+            f"(continue{instruction_counter})",
+        ]
+        return code, instruction_counter + 1
+    else:
+        raise ValueError(
+            f"Invalid command {stack_instruction.command!r} for stack instruction"
+        )
+
+
+def translate_memory_access_instruction(m: MemoryTransferCommand) -> list[str]:
```

Comment:
Rename to translate_memory_transfer_command. Change parameter name from m to command.